src/video/odk2/odk2.cc

```diff
@@ -0,0 +1,191 @@
+    m_OutputBufs.rawCamera = &m_RawCameraBuf;
+    m_OutputBufs.imu = &m_IMUBuf;
+    m_OutputBufs.state = &m_StateBuf;
+
+    // Clear atomic stop flag and start read thread
+    m_ShouldQuit = false;
+    m_ReadThread = std::thread(&ODK2::readThread, this);
+}
+//------------------------------------------------------------------------
+ODK2::~ODK2()
+{
+    // If thread is joinable, set quit flag and join
+    if(m_ReadThread.joinable()) {
+        m_ShouldQuit = true;
+        m_ReadThread.join();
+    }
+}
+//------------------------------------------------------------------------
+bool ODK2::readFrame(cv::Mat &outFrame)
+{
+    std::lock_guard<std::mutex> l(m_OutputBufsMutex);
+
+    // Ensure output image is correctly formatted
+    // **NOTE** we do this even if there's no image data so output frame is always right size
+    outFrame.create(BAND_HEIGHT, BAND_WIDTH, CV_8UC3);
+
+    // If there's a new frame
+    if(m_RawFrameBuf.timestampUs > m_LastRawFrameTimestep) {
+        using namespace units::literals;
+
+        // Update last timestep
+        m_LastRawFrameTimestep = m_RawFrameBuf.timestampUs;
+
+        // Extract quaternion components from frame
+        const float qw = m_RawFrameBuf.quaternion.w;
+        const float qx = m_RawFrameBuf.quaternion.x;
+        const float qy = m_RawFrameBuf.quaternion.y;
+        const float qz = m_RawFrameBuf.quaternion.z;
+
+        // Calculate heading from quaternion
+        const radian_t heading(std::atan2(2.0 * ((qw * qz) + (qx * qy)),
+                                          1.0 - (2.0 * ((qy * qy) + (qz * qz)))));
+        const uint32_t headingIdx = turn_t(heading + 180_deg).value() * BAND_WIDTH;
+
+        LOGD << "Processing frame heading=" << headingIdx;
+
+        // Loop through rows of horizontal band
+        for(int i = 0; i < BAND_HEIGHT; i++) {
+            // Get pointer to rows in outFrame and mask
+            // **NOTE** we want to flip output
+            cv::Vec3b *outPixel = outFrame.ptr<cv::Vec3b>(BAND_HEIGHT - i - 1);
+            uint8_t *outMask = m_Mask.ptr<uint8_t>(BAND_HEIGHT - i - 1);
+
+            // Loop through columns of horizontal band
+            for (int j = 0; j < BAND_WIDTH; j++) {
+                const int bandIdx = (i * BAND_WIDTH) + ((j + headingIdx) % BAND_WIDTH);
+
+                // Get fourpi index to map to band pixel
+                const int fourpiIdx = band_to_fourpi(bandIdx, bandType_horz);
+
+                // Create pixel into OpenCV image
+                const cv::Vec3b pixel(m_RawFrameBuf.image[fourpiIdx * 3],
+                                      m_RawFrameBuf.image[(fourpiIdx * 3) + 1],
+                                      m_RawFrameBuf.image[(fourpiIdx * 3) + 2]);
+
+                // Write pixel to OpenCV image
+                *outPixel++ = pixel;
+
+                // If pixel is black, clear mask otherwise set
+                *outMask++ = (pixel[0] == 0 && pixel[1] == 0 && pixel[2] == 0) ? 0 : 255;
```

Comment:
I'll stick my comments about the image masking here because it's as good a place as any!

Out of interest, are you currently using this mask somewhere? The reason I ask is that if the goal is to integrate it with the navigation code then I feel like it's not the right approach. One problem is that the navigation algos currently assume that the mask is set once (before training begins) and then remains unchanged, whereas I guess ``m_Mask`` will change with each frame (though probably only slightly).

I agree that it might be nice to add a virtual function to ``Video::Input`` for this case. How about ``loadMask()`` or something? It could return an ``ImgProc::Mask`` (not a reference) and the name sort of implies that an object is being created... In general my idea for the ``ImgProc::Mask`` class was that the underlying ``cv::Mat`` is mostly immutable: there's deliberately no way to get a non-const reference to it and if you want to update it you have to call ``set()`` again which ensures that a) it is of type ``CV_8UC1`` and b) all the values are either 0 or 255 (or it's empty). Then in theory we can just pass the object around as we like without having to do any checks where we're using it.

Reply:
I've just realised something: does the mask change on every frame because of the devkit's built-in image stabilisation? If so then this is probably a harder problem than I thought :grimacing:

Reply:
An additional point: even if we do want to derive a mask for each frame based on where the black pixels are I think it would make sense not to do it in this loop, to keep things simpler. There should be a relatively terse way to do this calculation after the fact using some vectorised OpenCV methods.

Reply:
It does change every frame because of the stabilisation hence doing this in the loop. I couldn't actually find a build in method which did it (``cv::compare`` was close but worked per-channel) - quite how ``cv::threshold`` worked with multi-channel wasn't clear...

Reply:
my other thought was to add a new ``set`` method to ``ImgProc::Mask`` which did whatever thresholding/other operation is required to make a mask (as that seems a reasonably general piece of functionality)

Reply:
(maybe ``cv::inRange`` can do this actually)

Reply:
I've had a thought about how to integrate this with the perfect memory code: we could just store a copy of the image mask with each stored snapshot instead of having a single mask for all of them. This is less awful than it sounds at first because ``cv::Mat``s are reference-counted and CoW so for the common case where the mask remains unchanged throughout training, all the stored masks would point to the same heap data.

Reply:
or we could make the perfect memory have the option of treating (0,0,0) as mask and handle it all automagically.

Reply:
I think from our previous discussions, as there's no biases, (0,0,0) already has no effect in InfoMax

Reply:
I think it's ok to treat the ``ImgProc::Mask`` as mutable in this case and come up with some way to fiddle with its internal data structure. I just wanted to make it hard to do this!

One way of doing it would be to derive a ``MutableMask`` class with a ``getMutable()`` method or something. Then we just make sure the return type of ``getMask()`` is the immutable ``ImgProc::Mask`` type.

Reply:
> or we could make the perfect memory have the option of treating (0,0,0) as mask and handle it all automagically.

Currently these methods only accept greyscale images so this would require some more hackery. I feel like being explicit about what we want our masks to be is probably the least error prone way of doing it. I've had enough trouble getting this explicit way to work correctly and I'd be worried if we have some implicit magic going on it'll make it really difficult to find bugs.

Reply:
eugh, that is a good point. It could be triggered by a ``CV_8UC2`` input type where the second channel is mask?

Reply:
that way, I think the change could be made at a pretty high level (in the perfect memory class itself)

Reply:
So we'd implement:

1. A flag in the ODK2 constructor which made ``ODK2::readGreyscaleFrame`` return ``CV_8UC2`` with gray + mask.
2. Make sure the raw store is capable of storing ``CV_8UC2``
3. Add asserts to ``PerfectMemory`` to prevent fixed masks and ``CV_8UC2`` images being used together.
4. Make a change whenever we call ``getMask()`` in ``PerfectMemory`` to use the second channel of the stored image/input image instead

Reply:
I've just installed the GitHub bot on Slack... let's see if it notices this message